### PR TITLE
Share ensemble smry datasets in memory between plugins

### PR DESCRIPTION
**Pull request by sigurdp**
This PR facilitates sharing of smry datasets between plugin instances in an effort to reduce the total memory consumption. 

Currently, each plugin instance that accesses smry data has its own copy of the data. For Webviz configurations involving several plugin instances and large datasets this results in a lot of memory overhead. Long term we will have to avoid loading whole smry datasets up front, and instead access data on demand in a much more fine grained way. Short term, this PR tries to alleviate the current situation by allowing smr data to be shared between plugin instances in many scenarios. 

The data sharing mechanism is realized through a global factory function for creating ensemble set models. When an `EnsembleSetModel` is created/retrieved through the new `caching_ensemble_set_model_factory.get_or_create_model()` function, the `ensemble_paths`, `time_index` and `column_keys` arguments will be combined and used a cache key so that plugin instances requesting the same ensemble set share the `EnsembleSetModel` instance. 

An important prerequisite for the approach taken by this PR is that all participating plugins must treat the shared smry data as read-only. The PR implements a simple check to try and detect if shared data has been tampered with, but the check is only performed when data is first accessed and is only intended as a "smoke-test" to try and detect unintentional modifications. 

Note that this PR implements sharing of the smry data at the ensemble set level. This approach is not the most optimal when the same ensemble ends up being accessed through multiple different ensemble sets. Lifting this restriction is possible, but will require significant refactoring of all plugins that utilize the data. Considering the large footprint of such a change and the likely upcoming redesign of the interfaces for data access, we decided to stick with this sharing level for the time being.

This PR partially resolved issue #545 as it reduces memory footprint for many Webviz usage scenarios.

Related PRs are #553, #556, #561 and #563.

---

### Contributor checklist

- [ ] :tada: This PR closes #ISSUE_NUMBER.
- [ ] :scroll: I have broken down my PR into the following tasks:
   - [ ] Task 1
   - [ ] Task 2
- [ ] :robot: I have added tests, or extended existing tests, to cover any new features or bugs fixed in this PR.
- [x] :book: I have considered adding a new entry in `CHANGELOG.md`, and added it if should be communicated there.

### CODE REVIEW

**Review by asnyv — COMMENTED** [review]
Just a couple of comments 🙂

File: webviz_subsurface/_models/caching_ensemble_set_model_factory.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    time_index: Optional[Union[list, str]] = None,
+    column_keys: Optional[list] = None,
+) -> EnsembleSetModel:
+
+    tmp_dict = {
+        "ensemble_paths": ensemble_paths,
+        "time_index": time_index,
+        "column_keys": column_keys,
+    }
+    modelkey = json.dumps(tmp_dict)
```
[the remark this answers is not in the document]
Comment (sigurdp) [reply webviz_subsurface/_models/caching_ensemble_set_model_factory.py:25]: Will fix, just a remnant from investigating the values using a print.

File: webviz_subsurface/_models/ensemble_set_model.py [hunk 2]
```diff
@@ -13,33 +13,39 @@ class EnsembleSetModel:
     def __init__(
         self,
         ensemble_paths: dict,
-        ensemble_set_name: str = "EnsembleSet",
-        filter_file: Union[str, None] = "OK",
```
[the remark this answers is not in the document]
Comment (sigurdp) [reply webviz_subsurface/_models/ensemble_set_model.py:17]: Was removed to keep the total number of arguments down now that we introduce two new ones. I suggest we re-introduce them when/if they are needed?

**Review by HansKallekleiv — APPROVED** [review]

**Comment by asnyv** [comment]
Btw @sigurdp: Can you add it to the changelog? Maybe also mention that you need to have the same ensembles defined in the different plugins to gain from the update?

**Comment by asnyv** [comment]
I started a rerun of the actions since it looked like it failed on one of these "random" errors.

**Pull request merged**